### ISSUE TITLE
margo-p2p-bw test hangs with high OS thread count (-T option)

### ISSUE THREAD

**Issue by carns** [mochi-hpc/mochi-margo#157]
Example: the following hangs on my laptop with na+sm and on spock with verbs:

`mpiexec -n 2 perf-regression/margo-p2p-bw -n na+sm -x 4096 -c 64 -D 10 -T 64`

That test program can be found in https://github.com/mochi-hpc-experiments/mochi-tests.  The hang occurs with `mochi-margo@main` and `argobots@main`, but not with `mochi-margo@0.9.5` and `argobots@main`.

As configured above, two processes will each allocate a 2 GiB memory region and then iterate through it with 4K bulk transfers for 10 seconds each using PUSH and PULL operations.  The -c 64 specifies that 64 ULTs will collaboratively work through the buffer, while the -T 64 indicates that those ULTs will execute on a pool with 64 execution streams (operating system threads).  The -c 64 configuration works fine as long as -T 64 is not also used.

The test program is a little dated in that it allocates the transfer pool explicitly rather than letting margo do it, but to my knowledge it isn't doing anything incorrectly.

This example with 64 OS threads competing for small transfers isn't necessarily going to perform _well_ (particularly on systems that don't have many CPU cores) but it should reliably complete without any functional problems.

**Comment by carns** [comment]
Confirmed that it has something to do either with Argobots eventual initializer or our usage of it.

If I build mochi-margo@main with argobots@main with this test case:
```
mpiexec -n 2 ./margo-p2p-bw -n na+sm -x 4096 -c 64 -D 1 -T 8
```
then the tests always either hangs or crashes with the following:
```
margo-p2p-bw: ../src/include/abti_waitlist.h:25: ABTI_waitlist_wait_and_unlock: Assertion `ABTD_spinlock_is_locked(p_lock) == 1' failed.
```

If I rebuild mochi-margo with `#undef ABT_EVENTUAL_INITIALIZER` at the top of margo-abt-macros.h then the same test case completes successfully.

**Comment by carns** [comment]
@mdorier I might need some help thinking through this; I keep confusing myself.

If abt eventuals are used directly (an `ABT_eventual`), then that type really a pointer to an internal type in Argobots, meaning that lines of code like the following are actually assigning a reference to the eventual, rather than a copy of the eventual:

https://github.com/mochi-hpc/mochi-margo/blob/main/src/margo-core.c#L886

In contrast, the Margo eventual implementation with static initialization has `margo_eventual_t` typedef'd to `ABT_eventual_memory`, which is a dummy struct (i.e., some opaque memory), not a reference:

https://github.com/mochi-hpc/mochi-margo/blob/main/src/margo-abt-macros.h#L26
(see https://github.com/pmodels/argobots/blob/main/src/include/abt.h.in#L1283)

That means that assignments like https://github.com/mochi-hpc/mochi-margo/blob/main/src/margo-core.c#L886 are making a copy of the eventual, rather than sharing a reference to it, right?

If so, that's problematic in a few places.

**Comment by carns** [comment]
Well, I modified that function to create the eventual directly in the req-> structure instead of creating one on the stack and assigning it, but the test case still doesn't work correctly.

Maybe a unit test is needed to see if we can reproduce in a simpler environment.

**Comment by carns** [comment]
WIP PR #158 has a rough unit test for margo eventuals that seems to work fine.  I'm not sure what's different.  Probably need to add some margo logging calls to the EVENTUAL macros (at least in an experimental build) and to the bw benchmark to try to figure out what sequence of events is causing  problem.

**Comment by carns** [comment]
The code path here inside of Argobots is straightforward.  Here is the relevant logic inside the `eventual_wait` call:

```
    ABTD_spinlock_acquire(&p_eventual->lock);
    if (p_eventual->ready == ABT_FALSE) {
        ABTI_waitlist_wait_and_unlock(&p_local, &p_eventual->waitlist,
                                      &p_eventual->lock,
                                      ABT_SYNC_EVENT_TYPE_EVENTUAL,
                                      (void *)p_eventual);
    } else {
        ABTD_spinlock_release(&p_eventual->lock);
    }
```
It is crashing in the `ABTI_waitlist_wait_and_unlock` call because a safety check (and value check in the debugger) indicates that `p_eventual->lock` is not acquired, but it was just acquired 2 lines up in eventual.c with `ABTD_spinlock_acquire`.

Possibly a problem with the atomic logic somewhere in here, but if so, I don't understand why this would trigger in this specific test case.  This code path is very heavily used with or without static eventuals.

**Comment by carns** [comment]
Maybe something in the atomic path doesn't work correctly if the static initializer is memcpy'd in instead of actually statically initialized on the spot?  Now that it's clear that the problem is triggering in the eventual for `margo_bulk_transfer` where the `margo_request_struct` is declared on the stack, we could try a hack that static initializes the eventual at that point to see if it makes a difference.

**Comment by carns** [comment]
The hack described in the previous comment looks like this:

```
    ABTD_spinlock_acquire(&p_eventual->lock);
    if (p_eventual->ready == ABT_FALSE) {
        ABTI_waitlist_wait_and_unlock(&p_local, &p_eventual->waitlist,
                                      &p_eventual->lock,
                                      ABT_SYNC_EVENT_TYPE_EVENTUAL,
                                      (void *)p_eventual);
    } else {
        ABTD_spinlock_release(&p_eventual->lock);
    }
```

... but makes no difference.  Going to open an issue on argobots repo for discussion to see if there are any other ideas for why we would hit that spinlock assertion.

**Comment by carns** [comment]
The hack described in the previous comment looks like this:

```
diff --git a/src/margo-core.c b/src/margo-core.c
index 2dc8e0f..34d71c5 100644
--- a/src/margo-core.c
+++ b/src/margo-core.c
@@ -1128,10 +1128,12 @@ static hg_return_t margo_bulk_itransfer_internal(
     margo_request     req) /* should have been allocated */
 {
     hg_return_t hret = HG_TIMEOUT;
+#if 0
     int         ret;
 
     ret = MARGO_EVENTUAL_CREATE(&(req->eventual));
     if (ret != 0) { return (HG_NOMEM_ERROR); }
+#endif
     req->timer          = NULL;
     req->handle         = HG_HANDLE_NULL;
     req->start_time     = ABT_get_wtime();
@@ -1153,7 +1155,7 @@ hg_return_t margo_bulk_transfer(margo_instance_id mid,
                                 size_t            local_offset,
                                 size_t            size)
 {
-    struct margo_request_struct reqs;
+    struct margo_request_struct reqs = {ABT_EVENTUAL_INITIALIZER, 0, NULL, NULL, 0, 0, 0, 0};
     hg_return_t                 hret = margo_bulk_itransfer_internal(
         mid, op, origin_addr, origin_handle, origin_offset, local_handle,
         local_offset, size, &reqs);
```

... but it doesn't make any difference.  I'll open an issue on the argobots repo for discussion of what could cause this case.

**Comment by carns** [comment]
Reproducer now available in #158 in the second test case.  It passes with no dedicated RPC execution streams, then hangs when there are two.

The key thing seems to be having the static eventual instantiated on a function stack that returns immediately after completing a wait() operation, with the function being called in a loop.

**Comment by carns** [comment]
See discussion on https://github.com/pmodels/argobots/issues/367

**Comment by carns** [comment]
It looks like the only safe way to use a static eventual is if you can insure that it will stay in scope until both the waiter and the setter are done, which we can't do in the margo blocking functions since the Mercury callbacks and the margo functions are decoupled.

We'll see if any ideas come out of the argobots github discussion.

**Comment by carns** [comment]
This was fixed by #161
